### ISSUE TITLE
close_primitive_builtin reverses type variables of operands

### ISSUE THREAD

**Issue by maz** [JHU-PL-Lab/big-bang#35]
The problem appears to occur when [`sequence`](https://github.com/maz/big-bang/blob/add-array/src/tiny-bang-typechecker/tiny_bang_constraint_closure.ml#L163) is called. I've tried testing `sequence` separately ([at line 28](https://github.com/maz/big-bang/blob/add-array/src/tiny-bang-typechecker/tiny_bang_constraint_closure.ml#L28)), but I can't get ocaml to believe that `sequence` exists.

@rpglover64 @leafac @zepalmer 

Based on code from https://github.com/maz/big-bang/blob/add-array/src/tiny-bang-typechecker/tiny_bang_constraint_closure.ml

```
iv = p1 \
     { p1 = ();
     } ->
     { r = ();
     };
length = 10;
arr = arrayNew length iv;
;;

BLARG operands
{length[],iv[]}
BLARG!!!
{p1 \ {p1 = ()} -> r*\{() <: r*},int}
{inconsistent!, int <: length[], p1 \ {p1 = ()} -> r*\{() <: r*} <: iv[], arrayNew length[] iv[] <: arr[]}

BLARG operands
{length[],iv[]}
BLARG!!!
{p1 \ {p1 = ()} -> r*\{() <: r*},int}
{inconsistent!, int <: length[], p1 \ {p1 = ()} -> r*\{() <: r*} <: iv[], arrayNew length[] iv[] <: arr[]}

Typechecking error discovered.
```

**Comment by leafac** [comment]
@maz I know this is going to sound dumb, but I have to ask because maybe it's going to help me understand the problem better: why don't just reverse the result in the end of the computation?

**Comment by maz** [comment]
Alex and I considered that. I'd like to first know why it's happening,
before saying that it will always be reversed.

**Comment by leafac** [comment]
Yes, of course! I'm glad to know my idea is not too off :smile:

**Comment by zepalmer** [comment]
I'm looking into this now.  Just so you know: in close_primitive_builtin, the sequence operation being used is the one on Nondeterminism_monad, which was locally opened within the function.

**Comment by zepalmer** [comment]
Here's some proof that it's the nondeterminism monad's fault.

```
zpalmer@thirtyseven:~$ rlwrap ocaml -init /dev/null
        OCaml version 4.02.2

# #use "topfind";;
- : unit = ()
Findlib has been successfully loaded. Additional directives:
  #require "package";;      to load a package
  #list;;                   to list the available packages
  #camlp4o;;                to load camlp4 (standard syntax)
  #camlp4r;;                to load camlp4 (revised syntax)
  #predicates "p,q,...";;   to set these predicates
  Topfind.reset();;         to force that packages will be reloaded
  #thread;;                 to enable threads

- : unit = ()
# #require "batteries";;
/home/zpalmer/.opam/4.02.2/lib/ocaml/unix.cma: loaded
/home/zpalmer/.opam/4.02.2/lib/ocaml/nums.cma: loaded
/home/zpalmer/.opam/4.02.2/lib/num-top: added to search path
/home/zpalmer/.opam/4.02.2/lib/num-top/num_top.cma: loaded
/home/zpalmer/.opam/4.02.2/lib/num: added to search path
/home/zpalmer/.opam/4.02.2/lib/ocaml/bigarray.cma: loaded
/home/zpalmer/.opam/4.02.2/lib/ocaml/str.cma: loaded
/home/zpalmer/.opam/4.02.2/lib/bytes: added to search path
/home/zpalmer/.opam/4.02.2/lib/batteries: added to search path
/home/zpalmer/.opam/4.02.2/lib/batteries/batteries.cma: loaded
/home/zpalmer/.opam/4.02.2/lib/batteries/batteriesConfig.cmo: loaded
/home/zpalmer/.opam/4.02.2/lib/batteries/batteriesHelp.cmo: loaded
/home/zpalmer/.opam/4.02.2/lib/batteries/batteriesPrint.cmo: loaded
# #require "monadlib";;
/home/zpalmer/.opam/4.02.2/lib/monadlib: added to search path
/home/zpalmer/.opam/4.02.2/lib/monadlib/monadlib.cma: loaded
# module Enum = Batteries.Enum;;
module Enum = Batteries.Enum
# module type Nondeterminism_monad_sig = sig
  type 'a m
  include Monad.LazyPlus with type 'a m := 'a m
  val pick_enum : 'a Enum.t -> 'a m
  val enum : 'a m -> 'a Enum.t

  val stop_unless : bool -> unit m
  end;;
module type Nondeterminism_monad_sig =
  sig
    type 'a m
    val bind : 'a m -> ('a -> 'b m) -> 'b m
    val return : 'a -> 'a m
    val ap : ('a -> 'b) m -> 'a m -> 'b m
    val lift1 : ('a -> 'b) -> 'a m -> 'b m
    val lift2 : ('a -> 'b -> 'c) -> 'a m -> 'b m -> 'c m
    val lift3 : ('a -> 'b -> 'c -> 'd) -> 'a m -> 'b m -> 'c m -> 'd m
    val sequence : 'a m list -> 'a list m
    val ( >>= ) : 'a m -> ('a -> 'b m) -> 'b m
    val join : 'a m m -> 'a m
    val zero : unit -> 'a m
    val plus : 'a m -> 'a m -> 'a m
    val null : 'a m -> bool
    val filter : ('a -> bool) -> 'a m -> 'a m
    val of_list : 'a list -> 'a m
    val mconcat : 'a m list -> 'a m
    val lplus : 'a m -> 'a m Lazy.t -> 'a m
    val of_llist : 'a LazyList.t -> 'a m
    val sum : 'a LazyList.t m -> 'a m
    val msum : 'a m LazyList.t -> 'a m
    val lsequence : 'a m LazyList.t -> 'a LazyList.t m
    val transpose : 'a LazyList.t m -> 'a m LazyList.t
    val pick_enum : 'a Enum.t -> 'a m
    val enum : 'a m -> 'a Enum.t
    val stop_unless : bool -> unit m
  end
# module Nondeterminism_monad_base = struct
  include Monad.LazyListM;;
  let enum m = LazyList.enum m;;
  end;;
module Nondeterminism_monad_base :
  sig
    type 'a m = 'a LazyList.t
    val bind : 'a m -> ('a -> 'b m) -> 'b m
    val return : 'a -> 'a m
    val zero : unit -> 'a m
    val lplus : 'a m -> 'a m Lazy.t -> 'a m
    val null : 'a m -> bool
    val enum : 'a LazyList.t -> 'a BatEnum.t
  end
# module Nondeterminism_monad : Nondeterminism_monad_sig = struct
  module M = Monad.MakeLazyPlus(Nondeterminism_monad_base);;
  include M;;
  let pick_enum e =
    let lazy_list_of_enum e =
      LazyList.unfold
        e
        (fun e' -> Batteries.Option.map (fun x -> (x,e')) @@ Enum.get e')
    in
    M.of_llist @@ lazy_list_of_enum e;;
  let enum = Nondeterminism_monad_base.enum;;

  let stop_unless x = if x then return () else zero ();;
  end;;
module Nondeterminism_monad : Nondeterminism_monad_sig
# open Batteries;;
# open Nondeterminism_monad;;
# let x = pick_enum @@ List.enum [1;2;3];;
val x : int Nondeterminism_monad.m = <abstr>
# let y = pick_enum @@ List.enum [4;5;6];;
val y : int Nondeterminism_monad.m = <abstr>
# List.of_enum @@ enum @@ sequence [x;y];;
- : int list list =
[[4; 1]; [5; 1]; [6; 1]; [4; 2]; [5; 2]; [6; 2]; [4; 3]; [5; 3]; [6; 3]]
# 
```

**Comment by zepalmer** [comment]
The cause of this problem is a bug in `monadlib` 0.1.  Unfortunately, the bugfix in 0.2 (which was tagged as a prerelease in the GitHub repository in January 2014) has not been released.

I  have adapted some code from the Odefa repository as a workaround.  I pushed the fix as d1a8ddf80e.  Please take a look and see if it works out for you.

**Comment by maz** [comment]
This works! See #36
